File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py

import sys

input = sys.stdin.readline

N, M = list(map(int, input().split()))

result = []

def backtrack(curr):
    if len(curr) == M:
        result.append(curr[:])
        return
    
    for i in range(1, N+1):
        if i not in curr:
            curr.append(i)
            backtrack(curr)
            curr.pop()

backtrack([])

for r in result:
    for c in r:
        print(c, end=" ")
    print()

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.swift
import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = input[0]
let m = input[1]

var visited: [Bool] = .init(repeating: false, count: n + 1)
var current: [Int] = []

func recursive(k: Int) {
  if k == m {
    // 출력
    for i in 0..<m {
      print("\(current[i])", terminator: " ")
    }
    print()
    return
  }

  for i in 1...n {
    if !visited[i] {
      visited[i] = true
      current.append(i)
      recursive(k: k + 1)
      current.removeLast()
      visited[i] = false
    }
  }
}

recursive(k: 0)